### PR TITLE
Add CSPDarkNet, DarkNet21 & DarkNet53

### PR DESCRIPTION
**Pull request by quantumalaviya**
# What does this PR do?
This PR implements the CSPDarkNet, DarkNet21, and DarkNet53 models commonly used with Yolo algorithms. Moreover, some of the building blocks are added as part of models/__internal__.

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [x] Did you write any new necessary tests?

## Who can review?

@LukeWood

### CODE REVIEW

**Comment by quantumalaviya** [comment]
This also fixes some of the links from my previous PR to comply with flake8. This time I made sure that the left bracket is on the same line as the text with the square brackets so there wouldn't be an issue while parsing the docstring.

File: keras_cv/models/__internal__/darknet_utils.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        )(x)
+        x = layers.BatchNormalization(name=f"{name}_bn")(x)
+
+        if activation == "silu":
+            x = layers.Lambda(
+                lambda x: keras.activations.swish(x), name=f"{name}_silu"
+            )(x)
+        elif activation == "relu":
+            x = layers.ReLU(name=f"{name}_relu")(x)
+        elif activation == "lrelu":
```
Comment (innat) [inline keras_cv/models/__internal__/darknet_utils.py:73]: Naming `leaky_relu` is more descriptive.

File: keras_cv/models/__internal__/darknet_utils.py [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+        self.batch_norm = layers.BatchNormalization()
+
+        if activation == "silu":
+            self.act = layers.Lambda(lambda x: keras.activations.swish(x))
+        elif activation == "relu":
+            self.act = layers.ReLU()
+        elif activation == "leaky_relu":
+            self.act = layers.LeakyReLU(0.1)
+
+    def call(self, x):
```
Comment (LukeWood) [inline keras_cv/models/__internal__/darknet_utils.py:73]: I think we can replace this entire block with a sequential model:

def DarknetConvBlock(layers.Layer):
    layers = [
        createConvAsYouDidAbove,  BatchNorm, activationAsYouDidAbove
    ]
    return keras.Sequential(layers)

File: keras_cv/models/__internal__/darknet_utils.py [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+                x = layers.Add(name=f"{name}_out")([residual, x])
+            else:
+                x = layers.Add(name=f"{name}_add_{i}")([residual, x])
+
+        return x
+
+    return apply
+
+
+def SPPBottleneck(
```
Comment (LukeWood) [inline keras_cv/models/__internal__/darknet_utils.py:144]: lets not abbreviate SPP

File: keras_cv/models/__internal__/darknet_utils.py [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+        x = DarknetConvBlock(
+            hidden_filters,
+            kernel_size=1,
+            strides=1,
+            activation=activation,
+            name=f"{name}_conv1",
+        )(x)
+        x = [x]
+
+        for kernel_size in kernel_sizes:
```
Comment (LukeWood) [inline keras_cv/models/__internal__/darknet_utils.py:179]: not an actionable comment, but what a cool architecture.  :)

File: keras_cv/models/__internal__/darknet_utils.py [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+            activation=activation,
+            name=f"{name}_conv2",
+        )(x)
+
+        return x
+
+    return apply
+
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
```
Comment (LukeWood) [inline keras_cv/models/__internal__/darknet_utils.py:203]: @fchollet @qlzh727 any issue with making these internal layers serializable?  I worry that this could break savedmodel creation in some way down the line if we need to change them.

File: keras_cv/models/__internal__/darknet_utils.py [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+
+    def __init__(self, filters, kernel_size, strides, activation="silu", **kwargs):
+        super().__init__(**kwargs)
+
+        self.filters = filters
+        self.kernel_size = kernel_size
+        self.strides = strides
+        self.activation = activation
+
+        self.dw_conv = layers.DepthwiseConv2D(
```
Comment (LukeWood) [inline keras_cv/models/__internal__/darknet_utils.py:228]: looks good, but same question as the above: would a sequential model for this actually be more idiomatic?

File: keras_cv/models/__internal__/darknet_utils.py [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+            "kernel_size": self.kernel_size,
+            "strides": self.strides,
+            "activation": self.activation,
+        }
+        base_config = super().get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class CSPLayer(layers.Layer):
```
Comment (LukeWood) [inline keras_cv/models/__internal__/darknet_utils.py:265]: lets remove abbreviations (CSP => CrossStagePartialLayer)

File: keras_cv/models/__internal__/darknet_utils.py [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+    layer converts the image into size (batch_size, width/2, height/2, 4*channels).
+
+    Args:
+        name: the name for the lambda layer used in the block.
+
+    Returns:
+        a function that takes an input Tensor representing a Focus layer.
+    """
+
+    def apply(x):
```
Comment (LukeWood) [inline keras_cv/models/__internal__/darknet_utils.py:384]: interesting, I wish I understood the purpose of this better. 

Maybe a docstring comment can serve to document this?

File: keras_cv/models/csp_darknet.py [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+        num_bottlenecks=base_depth,
+        add_residual=False,
+        use_depthwise=use_depthwise,
+        name="dark5_csp",
+    )(x)
+
+    if include_top:
+        x = layers.GlobalAveragePooling2D(name="avg_pool")(x)
+        x = layers.Dense(
+            num_classes, activation=classifier_activation, name="predictions"
```
Comment (LukeWood) [inline keras_cv/models/csp_darknet.py:166]: models look good to me overall

File: keras_cv/models/darknet.py [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+        x
+    )
+
+    # dark4
+    x = ResidualBlocks(filters=512, num_blocks=blocks[2], name="dark4_residual_block")(
+        x
+    )
+
+    # dark5
+    x = ResidualBlocks(filters=1024, num_blocks=blocks[3], name="dark5_residual_block")(
```
Comment (LukeWood) [inline keras_cv/models/darknet.py:161]: should this instead be `for filters, blocks in zip(filters, blocks)` with some rewriting to make this possible?

**Review by LukeWood — CHANGES_REQUESTED** [review]
minor changes

File: keras_cv/models/__init__.py [hunk 11]
```diff
@@ -12,6 +12,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from keras_cv.models.darknet import DarkNet21
+from keras_cv.models.darknet import DarkNet53
```
[the remark this answers is not in the document]
Comment (quantumalaviya) [reply keras_cv/models/__init__.py:16]: Yes, I must have missed it somewhere while merging with the darknet branch. I will update this PR soon.

File: keras_cv/models/__internal__/__init__.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from keras_cv.models.__internal__.darknet_utils import CSPLayer
+from keras_cv.models.__internal__.darknet_utils import DarknetConvBlock
+from keras_cv.models.__internal__.darknet_utils import DarknetConvBlockDepthwise
+from keras_cv.models.__internal__.darknet_utils import Focus
+from keras_cv.models.__internal__.darknet_utils import ResidualBlocks
+from keras_cv.models.__internal__.darknet_utils import SPPBottleneck
```
[the remark this answers is not in the document]
Comment (quantumalaviya) [reply keras_cv/models/__internal__/__init__.py:20]: It stands for Spatial Pyramid Pooling.

https://arxiv.org/pdf/1406.4729.pdf

File: keras_cv/models/__internal__/darknet_utils.py (hunk 2 above)
Comment (quantumalaviya) [reply keras_cv/models/__internal__/darknet_utils.py:73]: Can this cause serialization issues? That being said, I think it will definitely be more concise.

File: keras_cv/models/__internal__/darknet_utils.py (hunk 3 above)
Comment (quantumalaviya) [reply keras_cv/models/__internal__/darknet_utils.py:144]: `SpatialPyramidPoolingBottleneck` sounded a bit too long. I assumed this abbreviation made it more readable.

Should I change it to `SpatialPyramidPoolingBottleneck`?

File: keras_cv/models/__internal__/darknet_utils.py (hunk 8 above)
Comment (quantumalaviya) [reply keras_cv/models/__internal__/darknet_utils.py:384]: This is a new layer added with the YoloV5 implementation. Supposed to make the network more efficient.

https://github.com/ultralytics/yolov5/discussions/3181

File: keras_cv/models/darknet.py (hunk 10 above)
Comment (quantumalaviya) [reply keras_cv/models/darknet.py:161]: Should be doable, I'll try updating this.

File: keras_cv/models/__internal__/darknet_utils.py (hunk 3 above)
Comment (LukeWood) [reply keras_cv/models/__internal__/darknet_utils.py:144]: yeah, please do.  This is not a user facing API so a longer descriptive name is better

File: keras_cv/models/__internal__/darknet_utils.py (hunk 2 above)
Comment (LukeWood) [reply keras_cv/models/__internal__/darknet_utils.py:73]: no, Sequential models are layers, and models are also layers.  so serialization wise this is fine.

File: keras_cv/models/__internal__/darknet_utils.py (hunk 6 above)
Comment (LukeWood) [reply keras_cv/models/__internal__/darknet_utils.py:228]: please update to sequential models

File: keras_cv/models/__internal__/darknet_utils.py [hunk 13]
```diff
@@ -0,0 +1,399 @@
...
+            "kernel_size": self.kernel_size,
+            "strides": self.strides,
+            "activation": self.activation,
+        }
+        base_config = super().get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class CrossStagePartialLayer(layers.Layer):
```
Comment (LukeWood) [inline keras_cv/models/__internal__/darknet_utils.py:265]: rename CrossStagePartialLayer => CrossStagePartial

File: keras_cv/models/csp_darknet.py [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+    x = CrossStagePartialLayer(
+        base_channels * 8,
+        num_bottlenecks=base_depth * 3,
+        use_depthwise=use_depthwise,
+        name="dark4_csp",
+    )(x)
+
+    # dark5
+    x = ConvBlock(base_channels * 16, kernel_size=3, strides=2, name="dark5_conv")(x)
+    x = SPPBottleneck(
```
Comment (LukeWood) [inline keras_cv/models/csp_darknet.py:152]: SPP => SpatialPyramidPoolingBottleneck

**Review by LukeWood — CHANGES_REQUESTED** [review]
Minor changes

File: keras_cv/models/__internal__/__init__.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/models/__internal__/__init__.py:20]: Lets expand this acronym

File: keras_cv/models/__internal__/darknet_utils.py (hunk 2 above)
Comment (quantumalaviya) [reply keras_cv/models/__internal__/darknet_utils.py]: done!

**Review by LukeWood — APPROVED** [review]
Looks good to me

**Comment by LukeWood** [comment]
Thank you for your hard work @quantumalaviya

This is awesome, Merged
